feat(tracing): trace actions

dgozman opened the pull request:
- Fill and click actions pass metadata to Progress.
- Progress reports success/failure through instrumentation.
- Tracer consumes ActionResult and ActionMetadata and records them.

Currently, only click and fill actions pass metadata to
contain the size of the change. Everything else should follow.

yury-s commented on src/server/page.ts:
```diff
@@ -200,11 +199,7 @@ export class Page extends EventEmitter {
-        await agent.onBeforePageAction(this, progress);
-      return task(progress);
-    }, timeout);
+    return runAbortableTask(task, timeout);
```
Just in line this call?

dgozman replied:
Yeah, I'll follow up with that, to not make this patch even bigger.

yury-s commented on src/server/dom.ts:
```diff
@@ -367,11 +367,9 @@ export class ElementHandle<T extends Node = Node> extends js.JSHandle<T> {
   }
 
-  click(options: types.MouseClickOptions & types.PointerActionWaitOptions & types.NavigatingActionWaitOptions = {}): Promise<void> {
-    return this._page._runAbortableTask(async progress => {
```
Does it mean that it will now hang on the server side and be aborted only on the client? If so let's remove `timeout` from the options here.

dgozman replied:
Progress (that handles timeout) is now created in the ElementHandleDispatcher instead. Good point about the options, I will follow up with cleaning them up across the board.

yury-s commented on src/client/connection.ts:
```diff
@@ -71,12 +72,22 @@ export class Connection {
     const id = ++this._lastId;
     const validated = method === 'debugScopeState' ? params : validateParams(type, method, params);
-    const converted = { id, guid, method, params: validated };
+    const converted = { id, guid, method, params: validated, metadata: { stack } };
```
Why can't we manage metadata on the client and not send it over rpc?

dgozman replied:
This is a client-side stack that we want to see in the server-side tracing.

dgozman commented on src/client/connection.ts:
```diff
@@ -71,12 +72,22 @@ export class Connection {
     this.onmessage(converted);
-    return new Promise((resolve, reject) => this._callbacks.set(id, { resolve, reject }));
+    try {
+      const reuslt = await new Promise((resolve, reject) => this._callbacks.set(id, { resolve, reject }));
```
Done.

pavelfeldman commented on src/client/connection.ts:
```diff
@@ -71,12 +72,22 @@ export class Connection {
     this.onmessage(converted);
-    return new Promise((resolve, reject) => this._callbacks.set(id, { resolve, reject }));
+    try {
+      const reuslt = await new Promise((resolve, reject) => this._callbacks.set(id, { resolve, reject }));
```
reuslt

dgozman committed:
feat(tracing): trace actions

- Fill and click actions pass metadata to Progress.
- Progress reports success/failure through instrumentation.
- Tracer consumes ActionResult and ActionMetadata and records them.

Currently, only click and fill actions pass metadata to
contain the size of the change. Everything else should follow.
